### PR TITLE
Add support for shuffle reduction for the HIP backend

### PR DESCRIPTION
**Pull request by Rombur**
I haven't done any performance comparison yet but I have added a `FIXME_HIP_PERFORMANCE` comment so we don't forget to tune the algorithm.

### CODE REVIEW

File: core/src/HIP/Kokkos_HIP_Parallel_Range.hpp [hunk 1]
```diff
@@ -253,6 +259,49 @@ class ParallelReduce<FunctorType, Kokkos::RangePolicy<Traits...>, ReducerType,
     }
   }
 
+  __device__ inline void run(DummyShflReductionType) const {
+    value_type value;
+    ValueInit::init(ReducerConditional::select(m_functor, m_reducer), &value);
+    // Number of blocks is bounded so that the reduction can be limited to two
+    // passes. Each thread block is given an approximately equal amount of work
+    // to perform. Accumulate the values for this block. The accumulation
+    // ordering does not match the final pass, but is arithmatically equivalent.
```
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP_Parallel_Range.hpp:268]:
```suggestion
    // ordering does not match the final pass, but is arithmetically equivalent.
```

File: core/src/HIP/Kokkos_HIP_ReduceScan.hpp [hunk 2]
```diff
@@ -77,8 +77,15 @@ __device__ inline void hip_intra_warp_reduction(
         Kokkos::Experimental::shfl_down(result, blockDim.x * shift, warp_size);
     // Only join if upper thread is active (this allows non power of two for
     // blockDim.y)
-    if (threadIdx.y + shift < max_active_thread) join(result, tmp);
+    if (threadIdx.y + shift < max_active_thread) {
+      join(result, tmp);
+    }
     shift *= 2;
+    // Not sure why there is a race condition here but we need to wait for the
+    // join operation to be finished to perform the nex shuffle. Note that the
```
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP_ReduceScan.hpp:85]:
```suggestion
    // join operation to be finished to perform the next shuffle. Note that the
```

**Comment by masterleinad** [comment]
This looks OK to me comparing with the corresponding `CUDA` implementation and the other `HIP` implementation but I am not quite clear what the motivation for the pull request is.
AFAICT, it doesn't add any new functionality but should give better performance in the end?

**Comment by Rombur** [comment]
> AFAICT, it doesn't add any new functionality but should give better performance in the end?

Correct

File: core/src/HIP/Kokkos_HIP_ReduceScan.hpp [hunk 3]
```diff
@@ -191,18 +198,12 @@ __device__ bool hip_inter_block_reduction(
...
-      // FIXME_HIP check that the ballot is necessary. Also active is not read.
-      int active = __ballot(1);
-      for (unsigned int i = 2; i < warp_size; i *= 2) {
+      for (unsigned int i = 1; i < warp_size; i *= 2) {
         if ((blockDim.x * blockDim.y) > i) {
           value_type tmp = Kokkos::Experimental::shfl_down(value, i, warp_size);
           if (id + i < gridDim.x) join(value, tmp);
         }
-        active += __ballot(1);
+        __syncthreads();
```
Comment (crtrott) [inline core/src/HIP/Kokkos_HIP_ReduceScan.hpp:206]: wait why do you do the syncthreads here? THis is within a warp?

File: core/src/HIP/Kokkos_HIP_ReduceScan.hpp [hunk 4]
```diff
@@ -231,6 +232,7 @@ __device__ inline void hip_intra_warp_reduction(
     // blockDim.y)
     if (threadIdx.y + shift < max_active_thread) reducer.join(result, tmp);
     shift *= 2;
+    __syncthreads();
```
Comment (crtrott) [inline core/src/HIP/Kokkos_HIP_ReduceScan.hpp:235]: again why syncthreads?

File: core/src/HIP/Kokkos_HIP_ReduceScan.hpp [hunk 5]
```diff
@@ -351,18 +353,12 @@ __device__ inline bool hip_inter_block_reduction(
...
-      // FIXME_HIP check that the ballot is necessary. Also active is not read.
-      int active = __ballot(1);
-      for (unsigned int i = 2; i < 64; i *= 2) {
+      for (unsigned int i = 1; i < warp_size; i *= 2) {
         if ((blockDim.x * blockDim.y) > i) {
           value_type tmp = Kokkos::Experimental::shfl_down(value, i, warp_size);
           if (id + i < gridDim.x) reducer.join(value, tmp);
         }
-        active += __ballot(1);
+        __syncthreads();
```
Comment (crtrott) [inline core/src/HIP/Kokkos_HIP_ReduceScan.hpp:361]: why syncthreads?

**Comment by Rombur** [comment]
> There are syncthreads in the warp level reductions. Why?

Because I have a race condition at the warp level :( I plan to bring the problem with AMD at our next meeting with them. I want to point out that there is a similar problem with CUDA Clang and there you fixed it by calling dummy `__ballot(1)` but that doesn't work for me. See https://github.com/kokkos/kokkos/issues/941

**Pull request merged**